pkg/cvo/cvo_integration_test.go

```diff
@@ -0,0 +1,712 @@
+		t.Fatal(err)
+	}
+	payloadImage1 := "arbitrary/release:image"
+	payloadImage2 := "arbitrary/release:image-2"
+	retriever := &mapPayloadRetriever{map[string]string{
+		payloadImage1: filepath.Join(dir, "0.0.1"),
+		payloadImage2: filepath.Join(dir, "0.0.2"),
+	}}
+
+	optr := New(
+		"", ns, ns, payloadImage1,
+		filepath.Join(dir, "ignored"),
+		5*time.Second,
+		cvInformer.Config().V1().ClusterVersions(),
+		sharedInformers.Config().V1().ClusterOperators(),
+		cfg,
+		client, kc, apiExtClient,
+		false,
+	)
+
+	worker := optr.configSync.(*SyncWorker)
+	worker.retriever = retriever
+
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	go cvInformer.Start(stopCh)
+	go sharedInformers.Start(stopCh)
+	go optr.Run(1, stopCh)
+
+	t.Logf("wait until we observe the cluster version become available")
+	lastCV, err := waitForAvailableUpdate(t, client, ns, false, "0.0.1")
+	if err != nil {
+		t.Logf("latest version:\n%s", printCV(lastCV))
+		t.Fatalf("cluster version never became available: %v", err)
+	}
+
+	status := optr.configSync.(*SyncWorker).Status()
+
+	t.Logf("verify the available cluster version's status matches our expectations")
+	t.Logf("Cluster version:\n%s", printCV(lastCV))
+	verifyClusterVersionStatus(t, lastCV, configv1.Update{Payload: payloadImage1, Version: "0.0.1"}, 1)
+	verifyReleasePayload(t, kc, ns, "0.0.1", payloadImage1)
+
+	t.Logf("wait for the next resync and verify that status didn't change")
+	if err := wait.Poll(time.Second, 30*time.Second, func() (bool, error) {
+		updated := optr.configSync.(*SyncWorker).Status()
+		if updated.Completed > status.Completed {
+			return true, nil
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+	time.Sleep(time.Second)
+	cv, err := client.Config().ClusterVersions().Get(ns, metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !reflect.DeepEqual(cv.Status, lastCV.Status) {
+		t.Fatalf("unexpected: %s", diff.ObjectReflectDiff(lastCV.Status, cv.Status))
+	}
+	verifyReleasePayload(t, kc, ns, "0.0.1", payloadImage1)
+
+	t.Logf("trigger an update to a new version")
+	cv, err = client.Config().ClusterVersions().Patch(ns, types.MergePatchType, []byte(fmt.Sprintf(`{"spec":{"desiredUpdate":{"payload":"%s"}}}`, payloadImage2)))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if cv.Spec.DesiredUpdate == nil {
+		t.Fatalf("cluster desired version was not preserved: %s", printCV(cv))
+	}
+
+	t.Logf("wait for the new version to be available")
+	lastCV, err = waitForAvailableUpdate(t, client, ns, false, "0.0.1", "0.0.2")
+	if err != nil {
+		t.Logf("latest version:\n%s", printCV(lastCV))
+		t.Fatalf("cluster version never reached available at 0.0.2: %v", err)
+	}
+	t.Logf("Upgraded version:\n%s", printCV(lastCV))
+	verifyClusterVersionStatus(t, lastCV, configv1.Update{Payload: payloadImage2, Version: "0.0.2"}, 2)
+	verifyReleasePayload(t, kc, ns, "0.0.2", payloadImage2)
+
+	t.Logf("delete an object so that the next resync will recover it")
```

Comment:
This is fine for now, but we should move this to separate test if we can later on.

Reply:
so the integration tests are still sort of scenario driven, because it does take non-zero time to do the payload.  I'm halfway in between granular and full scenario tests here.